### DISCUSSION TITLE
Native window control (DWM, NS, GTK) + Plugin Architecture

### DISCUSSION THREAD

**Discussion by brys0** [wailsapp/wails#1587] [Ideas]
I'm not sure if this exists yet but I couldn't find any documentation on the topic. I'd appreciate it if we had the ability to add our own methods or controllers for the window. For example using experimental unreleased DWM flags from Windows would be excellent. Having the ability to also use custom items from NS for Mac and GTK for Linux would be amazing. Following this, the ability to create wails plugins (maybe it call it wailins) would be great for people who don't want to dive head first into golang for everything.

**Comment by leaanthony** [comment]
Hi 👋 Any thoughts on what this would look like or how it might work?

**Reply by brys0** [reply]
Hi! Thanks for the quick response back, yes I do. 
## For the plugins:

### In the Application before we use `Wails.run` you could have the option to register a plugin which contains the following function callbacks:

`initialize()` - Ran when Wails.run finishes.
`initalizeScript()` - Returns a string (Optional) which provides the JS script to be executed on the webview
`name()` - Gets the name of the plugin i.e 'My fav plugin' (This could also be a default struct)
`created()` - Ran when the window is created 
`onPageLoad()` - Ran when the webview has a navigation  

Each of these functions should have access to the App class/runtime and the config file for custom options to control in the config.json (Or whatever it is you use)
## For native window control:

Have the ability to 'extend' native window control beyond what is currently available. 
As far as my last encounter/experience with wails beta I was having trouble adding a custom DWM type that allows MICA for win 11+ devices. There was no way to directly change that. (That I could find) this feature would allow (optionally) each software target [Linux,Windows,MacOS] to have a option or callback for working with the native window and adding custom flags or options to it. Beyond what you allow.

**Reply by leaanthony** [reply]
Thanks for your thoughts. I'm not against this at all, but there are some things to consider. For instance, on Windows, (I think) this would mandate the need for requiring CGO? If we're just talking about hooks so that the developer can have custom window creation code then that wouldn't be too much of an issue (I think).

**Reply by brys0** [reply]
Ideally the native windows creation would just be able to hook into the creation process. However, I don't think that would be possible without CGO. It's up to you really. Just a thought 👍🏻
